app/javascript/mastodon/features/explore/results.jsx

```diff
@@ -49,60 +52,128 @@ const renderStatuses = (results, onLoadMore) => appendLoadMore('statuses', resul
 class Results extends PureComponent {
 
   static propTypes = {
-    results: ImmutablePropTypes.map,
+    results: ImmutablePropTypes.contains({
+      accounts: ImmutablePropTypes.orderedSet,
+      statuses: ImmutablePropTypes.orderedSet,
+      hashtags: ImmutablePropTypes.orderedSet,
+    }),
     isLoading: PropTypes.bool,
     multiColumn: PropTypes.bool,
     dispatch: PropTypes.func.isRequired,
     q: PropTypes.string,
     intl: PropTypes.object,
+    submittedType: PropTypes.oneOf(['accounts', 'statuses', 'hashtags']),
   };
 
   state = {
-    type: 'all',
+    type: this.props.submittedType || 'all',
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    if (props.submittedType !== state.type) {
+      return {
+        type: props.submittedType || 'all',
+      };
+    }
+
+    return null;
+  };
+
+  handleSelectAll = () => {
+    const { submittedType, dispatch } = this.props;
+
+    // If we originally searched for a specific type, we need to resubmit
+    // the query to get all types of results
+    if (submittedType) {
+      dispatch(submitSearch());
+    }
+
+    this.setState({ type: 'all' });
   };
 
-  handleSelectAll = () => this.setState({ type: 'all' });
-  handleSelectAccounts = () => this.setState({ type: 'accounts' });
-  handleSelectHashtags = () => this.setState({ type: 'hashtags' });
-  handleSelectStatuses = () => this.setState({ type: 'statuses' });
-  handleLoadMoreAccounts = () => this.loadMore('accounts');
-  handleLoadMoreStatuses = () => this.loadMore('statuses');
-  handleLoadMoreHashtags = () => this.loadMore('hashtags');
+  handleSelectAccounts = () => {
+    const { submittedType, dispatch } = this.props;
+
+    // If we originally searched for something else (but not everything),
+    // we need to resubmit the query for this specific type
+    if (!!submittedType && submittedType !== 'accounts') {
+      dispatch(submitSearch('accounts'));
+    }
```

Comment:
In conjunction with `getDerivedStateFromProps`, it means it's not possible to switch from the “All” tab, as `submittedType` will be unset, and `type` will revert to `all`.